### PR TITLE
ui: display conferences between dates

### PR DESCRIPTION
**Pull request by miguelgrc**
* INSPIR-2999

### CODE REVIEW

File: ui/src/submissions/conferences/components/ExistingConferencesDrawer.scss [hunk 1]
```diff
@@ -0,0 +1,17 @@
+.__ExistingConferencesDrawer__ {
+  .ant-drawer-content-wrapper {
+    width: 25% !important;
+    padding-top: $header-min-height;
```
[the remark this answers is not in the document]
Comment (harunurhan) [reply ui/src/submissions/conferences/components/ExistingConferencesDrawer.scss]: if we really need to display this inside "content" part of our layout, you should use `getContainer` prop of `Drawer`

File: ui/src/submissions/conferences/containers/ExistingConferencesAlertContainer.jsx [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+
+export function ExistingConferencesAlert({
+  dispatch,
+  dates,
+  numberOfConferences,
+}) {
+  const [openingDate, closingDate] = dates;
+  const [showDrawer, setShowDrawer] = useState(false);
+
+  useMemo(
```
Comment (harunurhan) [inline ui/src/submissions/conferences/containers/ExistingConferencesAlertContainer.jsx]: `useMemo`, memoizing "expensive" operations to improve performance, you should use `useEffect` for side effects like dispatching actions.

File: ui/src/submissions/conferences/containers/ExistingConferencesAlertContainer.jsx [hunk 3]
```diff
@@ -0,0 +1,85 @@
...
+} from '../../../common/constants';
+import { searchQueryUpdate } from '../../../actions/search';
+
+export function ExistingConferencesAlert({
+  dispatch,
+  dates,
+  numberOfConferences,
+}) {
+  const [openingDate, closingDate] = dates;
+  const [showDrawer, setShowDrawer] = useState(false);
```
Comment (harunurhan) [inline ui/src/submissions/conferences/containers/ExistingConferencesAlertContainer.jsx]: `isDrawerVisible`, `setDrawerVisible` to be consistent with naming conventions

File: ui/src/submissions/conferences/containers/ExistingConferencesAlertContainer.jsx [hunk 4]
```diff
@@ -0,0 +1,85 @@
...
+
+  useMemo(
+    () => {
+      if (openingDate && closingDate) {
+        dispatch(
+          searchQueryUpdate(EXISTING_CONFERENCES_NS, {
+            contains: `${openingDate}${RANGE_AGGREGATION_SELECTION_SEPARATOR}${closingDate}`,
+            start_date: START_DATE_ALL,
```
Comment (harunurhan) [inline ui/src/submissions/conferences/containers/ExistingConferencesAlertContainer.jsx]: you can remove this, after you set in the `baseQuery` and `query`

File: ui/src/submissions/conferences/containers/ExistingConferencesAlertContainer.jsx [hunk 5]
```diff
@@ -0,0 +1,85 @@
...
+                onClick={() => setShowDrawer(true)}
+              >
+                Show more
+              </Button>
+            </span>
+          }
+        />
+        <ExistingConferencesDrawer
+          visible={showDrawer}
+          onClose={() => setShowDrawer(false)}
```
Comment (harunurhan) [inline ui/src/submissions/conferences/containers/ExistingConferencesAlertContainer.jsx]: `onDrawerClose = useCallback(() => ...)`

File: ui/src/submissions/conferences/containers/ExistingConferencesAlertContainer.jsx [hunk 6]
```diff
@@ -0,0 +1,85 @@
...
+
+const stateToProps = state => ({
+  numberOfConferences: state.search.getIn([
+    'namespaces',
+    EXISTING_CONFERENCES_NS,
+    'total',
+  ]),
+});
+
+const dispatchToProps = dispatch => ({ dispatch });
```
Comment (harunurhan) [inline ui/src/submissions/conferences/containers/ExistingConferencesAlertContainer.jsx]: instead of pass the `dispatch` you should pass a prop `onDatesChange` that dispatches the action then call this in the component (see other example containers).

File: ui/src/submissions/conferences/containers/ExistingConferencesAlertContainer.jsx [hunk 7]
```diff
@@ -0,0 +1,85 @@
...
+import pluralizeUnlessSingle from '../../../common/utils';
+import ExistingConferencesDrawer from '../components/ExistingConferencesDrawer';
+import { EXISTING_CONFERENCES_NS } from '../../../reducers/search';
+import {
+  RANGE_AGGREGATION_SELECTION_SEPARATOR,
+  START_DATE_ALL,
+} from '../../../common/constants';
+import { searchQueryUpdate } from '../../../actions/search';
+
+export function ExistingConferencesAlert({
```
Comment (harunurhan) [inline ui/src/submissions/conferences/containers/ExistingConferencesAlertContainer.jsx]: you should test (snapshots) the component, perhaps separate to another file.

File: ui/src/submissions/conferences/components/ExistingConferencesDrawer.jsx [hunk 8]
```diff
@@ -0,0 +1,44 @@
...
+import ConferenceItem from '../../../conferences/components/ConferenceItem';
+import { EXISTING_CONFERENCES_NS } from '../../../reducers/search';
+import PaginationContainer from '../../../common/containers/PaginationContainer';
+import pluralizeUnlessSingle from '../../../common/utils';
+
+function renderConferenceItem(result) {
+  return <ConferenceItem metadata={result.get('metadata')} />;
+}
+
+function ExistingConferencesDrawer({ visible, onClose, numberOfConferences }) {
```
Comment (harunurhan) [inline ui/src/submissions/conferences/components/ExistingConferencesDrawer.jsx]: you should test this one too

File: ui/package.json [hunk 9]
```diff
@@ -2,7 +2,7 @@
   "name": "inspire-next-react",
   "license": "GPL-2.0",
   "version": "0.1.19",
-  "proxy": "http://localhost:8000",
+  "proxy": "http://localhost:8080",
```
Comment (harunurhan) [inline ui/package.json]: left over

File: ui/src/submissions/conferences/containers/ExistingConferencesAlertContainer.jsx (hunk 5 above)
Comment (miguelgrc) [reply ui/src/submissions/conferences/containers/ExistingConferencesAlertContainer.jsx]: Why the `useCallback` here?

File: ui/src/submissions/conferences/components/__tests__/ExistingConferencesAlert.test.jsx [hunk 10]
```diff
@@ -0,0 +1,21 @@
+import React from 'react';
+import { shallow } from 'enzyme';
+
+import ExistingConferencesAlert from '../ExistingConferencesAlert';
+
+describe('ExistingConferencesAlert', () => {
```
Comment (harunurhan) [inline ui/src/submissions/conferences/components/__tests__/ExistingConferencesAlert.test.jsx]: we need to test `onDatesChange `. However,  for now there is no straight forward way to test `useEffect`, when you have a component that has containers in the subtree. Because you need a `mount`, which requires you to wrap the component with `Provider`, then you can use `wrapper.setProps` because it would reference to `Provider`.

TLDR; put a  comment ```// TODO: test onDatesChange when issue of testing useEffect with shallow rendering is resolved```, I will do little investigation and we'll add those after.

File: ui/src/submissions/conferences/containers/__tests__/ExistingConferencesAlertContainer.test.jsx [hunk 11]
```diff
@@ -0,0 +1,37 @@
...
+import { Provider } from 'react-redux';
+
+import ExistingConferencesAlertContainer from '../ExistingConferencesAlertContainer';
+import { EXISTING_CONFERENCES_NS } from '../../../../reducers/search';
+import { SEARCH_QUERY_UPDATE } from '../../../../actions/actionTypes';
+import { getStore } from '../../../../fixtures/store';
+import { RANGE_AGGREGATION_SELECTION_SEPARATOR } from '../../../../common/constants';
+
+describe('ExistingConferencesAlertContainer', () => {
+  it('dispatches SEARCH_QUERY_UPDATE', () => {
```
Comment (harunurhan) [inline ui/src/submissions/conferences/containers/__tests__/ExistingConferencesAlertContainer.test.jsx]: `dispatches SEARCH_QUERY_UPDATE initially`

we have to `when dates is changed` but that's also tricky because of a reason I mentioned above. Leave a similar `TODO` for now :/, Perhaps our approach is odd or possibly wrong, that's why it becomes harder to test. I'll check tomorrow, how we can do this.

File: ui/src/submissions/conferences/components/ExistingConferencesAlert.jsx [hunk 12]
```diff
@@ -0,0 +1,61 @@
...
+              found in these dates.
+              <Button type="link" size="small" onClick={() => showDrawer(true)}>
+                Show more
+              </Button>
+            </span>
+          }
+        />
+        <ExistingConferencesDrawer
+          visible={isDrawerVisible}
+          onDrawerClose={() => showDrawer(false)}
```
Comment (harunurhan) [inline ui/src/submissions/conferences/components/ExistingConferencesAlert.jsx]: you have to set memoized callback (what is returned from `userCallback`) directly as a prop, without a wrapper arrow function.

File: ui/src/submissions/conferences/components/ExistingConferencesAlert.jsx [hunk 13]
```diff
@@ -0,0 +1,61 @@
...
+
+function ExistingConferencesAlert({
+  onDatesChange,
+  dates,
+  numberOfConferences,
+}) {
+  const [openingDate, closingDate] = dates;
+  const [isDrawerVisible, setDrawerVisible] = useState(false);
+
+  const showDrawer = useCallback(visible => setDrawerVisible(visible), []);
```
Comment (harunurhan) [inline ui/src/submissions/conferences/components/ExistingConferencesAlert.jsx]: `const onDrawerClose = useCallback(() => setDrawerVisible(false), []);`

File: ui/src/submissions/conferences/components/ExistingConferencesAlert.jsx [hunk 14]
```diff
@@ -0,0 +1,61 @@
...
+
+  useEffect(
+    () => {
+      if (openingDate && closingDate) {
+        onDatesChange(openingDate, closingDate);
```
Comment (harunurhan) [inline ui/src/submissions/conferences/components/ExistingConferencesAlert.jsx]: `onDatesChange([openingDate, closingDate])` to be consistent with how we receive them.

File: ui/src/submissions/conferences/components/ExistingConferencesAlert.jsx [hunk 15]
```diff
@@ -0,0 +1,62 @@
...
+
+function ExistingConferencesAlert({
+  onDatesChange,
+  dates,
+  numberOfConferences,
+}) {
+  const [openingDate, closingDate] = dates;
+  const [isDrawerVisible, setDrawerVisible] = useState(false);
+
+  const onDrawerOpen = useCallback(() => setDrawerVisible(true), []);
```
Comment (harunurhan) [inline ui/src/submissions/conferences/components/ExistingConferencesAlert.jsx]: `onShowMoreClick`

File: ui/src/submissions/conferences/containers/__tests__/ExistingConferencesAlertContainer.test.jsx [hunk 16]
```diff
@@ -0,0 +1,39 @@
...
+          query: {
+            contains: `${startingDate}${RANGE_AGGREGATION_SELECTION_SEPARATOR}${closingDate}`,
+          },
+        },
+      },
+    ];
+    expect(store.getActions()).toEqual(expectedActions);
+  });
+
+  // TODO: Test when dates change when issue of testing useEffect with shallow rendering is resolved
```
Comment (harunurhan) [inline ui/src/submissions/conferences/containers/__tests__/ExistingConferencesAlertContainer.test.jsx]: you can remove this

File: ui/src/submissions/conferences/containers/__tests__/ExistingConferencesAlertContainer.test.jsx [hunk 17]
```diff
@@ -0,0 +1,39 @@
...
+import { Provider } from 'react-redux';
+
+import ExistingConferencesAlertContainer from '../ExistingConferencesAlertContainer';
+import { EXISTING_CONFERENCES_NS } from '../../../../reducers/search';
+import { SEARCH_QUERY_UPDATE } from '../../../../actions/actionTypes';
+import { getStore } from '../../../../fixtures/store';
+import { RANGE_AGGREGATION_SELECTION_SEPARATOR } from '../../../../common/constants';
+
+describe('ExistingConferencesAlertContainer', () => {
+  it('dispatches SEARCH_QUERY_UPDATE initially', () => {
```
Comment (harunurhan) [inline ui/src/submissions/conferences/containers/__tests__/ExistingConferencesAlertContainer.test.jsx]: and test onDatesChange logic only. An example for a different container:

```jsx
  it('dispatches CITATIONS_SUMMARY_REQUEST on CitationSummaryBox query change', () => {
    const initialQuery = { sort: 'mostcited', q: 'cern' };
    const store = getStoreWithState({
      search: fromJS({
        namespaces: {
          [LITERATURE_NS]: {
            query: initialQuery,
          },
        },
      }),
    });
    const wrapper = mount(
      <Provider store={store}>
        <CitationSummaryBoxContainer />
      </Provider>
    );
    store.clearActions(); // to clear initial CITATIONS_SUMMARY_REQUEST that is dispatched on mount

    const newQuery = { experment: 'atlas' };
    const onQueryChange = wrapper
      .find(CitationSummaryBox)
      .prop('onQueryChange');
    onQueryChange(newQuery);

    const expectedActions = [
      {
        payload: { query: newQuery },
        type: CITATIONS_SUMMARY_REQUEST,
      },
    ];
    expect(store.getActions()).toEqual(expectedActions);
  });
```

File: ui/src/submissions/conferences/containers/__tests__/ExistingConferencesAlertContainer.test.jsx [hunk 18]
```diff
@@ -0,0 +1,39 @@
...
+import { mount } from 'enzyme';
+import { Provider } from 'react-redux';
+
+import ExistingConferencesAlertContainer from '../ExistingConferencesAlertContainer';
+import { EXISTING_CONFERENCES_NS } from '../../../../reducers/search';
+import { SEARCH_QUERY_UPDATE } from '../../../../actions/actionTypes';
+import { getStore } from '../../../../fixtures/store';
+import { RANGE_AGGREGATION_SELECTION_SEPARATOR } from '../../../../common/constants';
+
+describe('ExistingConferencesAlertContainer', () => {
```
Comment (harunurhan) [inline ui/src/submissions/conferences/containers/__tests__/ExistingConferencesAlertContainer.test.jsx]: you need to test that `numberOfConferences are passed to the component properly. An example for a different container:

```jsx
  it('passes literature query', () => {
    const store = getStoreWithState({
      search: fromJS({
        namespaces: {
          [LITERATURE_NS]: {
            query: { sort: 'mostcited', q: 'query' },
          },
        },
      }),
    });
    const wrapper = mount(
      <Provider store={store}>
        <CitationSummaryBoxContainer />
      </Provider>
    );
    expect(wrapper.find(CitationSummaryBox)).toHaveProp({
      query: { sort: 'mostcited', q: 'query' },
    });
  });
```
